### PR TITLE
MoreVP - Exit with burn in-flight

### PR DESCRIPTION
**Pull request by atvanguard**
Exit from a burn in-flight tx by referencing an older tx.
https://github.com/maticnetwork/contracts/issues/48

### CODE REVIEW

File: contracts/root/lib/ExitTxValidator.sol [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    require(
+      childToken == RLPReader.toAddress(txList[3]), // corresponds to "to" field in tx)
+      "Reference and exit tx do not correspond to the same token"
+    );
+    bytes memory txData = RLPReader.toBytes(txList[5]);
+    bytes4 funcSig = BytesLib.toBytes4(BytesLib.slice(txData, 0, 4));
+    if (funcSig == WITHDRAW_FUNC_SIG) {
+      require(txData.length == 36, "Invalid tx"); // 4 bytes for funcSig and a single bytes32 parameter
+      exitAmountOrTokenId = BytesLib.toUint(txData, 4);
+      burnt = true;
```
Comment (jdkanani) [inline contracts/root/lib/ExitTxValidator.sol]: Make sure we don't set burnt true for every transaction.

File: contracts/root/withdrawManager/WithdrawManager.sol [hunk 2]
```diff
@@ -31,6 +32,73 @@ contract WithdrawManager is WithdrawManagerStorage /* , IWithdrawManager */ {
...
+      address(registry)
+    );
+    require(
+      registry.rootToChildToken(rootToken) == childToken,
+      "INVALID_ROOT_TO_CHILD_TOKEN_MAPPING"
+    );
+
+    // validate exitTx
+    (uint256 exitAmountOrTokenId, bool burnt) = ExitTxValidator.processExitTx(exitTx, childToken, msg.sender);
+    // gives stack too deep, fix later
```
Comment (jdkanani) [inline contracts/root/withdrawManager/WithdrawManager.sol]: Fix stack too deep :)

File: contracts/root/withdrawManager/WithdrawManager.sol [hunk 3]
```diff
@@ -237,8 +305,8 @@ contract WithdrawManager is WithdrawManagerStorage /* , IWithdrawManager */ {
     uint256 _exitId = (
       headerNumber * HEADER_BLOCK_NUMBER_WEIGHT +
       withdrawBlockNumber * WITHDRAW_BLOCK_NUMBER_WEIGHT +
-      path.toRlpItem().toBytes().toRlpItem().toUint() * 100000 +
-      oIndex
+      path.toRlpItem().toBytes().toRlpItem().toUint() * 100000
+      // + oIndex
```
Comment (jdkanani) [inline contracts/root/withdrawManager/WithdrawManager.sol]: Will exit Ids be the same?

File: contracts/root/withdrawManager/WithdrawManager.sol (hunk 2 above)
Comment (atvanguard) [reply contracts/root/withdrawManager/WithdrawManager.sol]: Have fixed this along with some changes in the next PR.

File: contracts/root/lib/ExitTxValidator.sol (hunk 1 above)
Comment (atvanguard) [reply contracts/root/lib/ExitTxValidator.sol]: For sure :)

File: contracts/root/withdrawManager/WithdrawManager.sol (hunk 3 above)
Comment (atvanguard) [reply contracts/root/withdrawManager/WithdrawManager.sol]: Yes, the way we calculate the *age* will remain same like in MVL. Just that we will take the age of reference tx now. I'll probably give an exit ID that'll be different from *age*, like sequential maybe - Yet to think about the *implementation details* of the challenges, so using age right now.

**Pull request merged**